src/main/kotlin/eu/theblob42/idea/whichkey/WhichKeyActionListener.kt

```diff
@@ -0,0 +1,114 @@
+package eu.theblob42.idea.whichkey
+
+import com.intellij.openapi.actionSystem.*
+import com.intellij.openapi.actionSystem.ex.AnActionListener
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.wm.WindowManager
+import com.maddyhome.idea.vim.api.injector
+import com.maddyhome.idea.vim.command.Argument
+import com.maddyhome.idea.vim.command.CommandState
+import com.maddyhome.idea.vim.command.MappingMode
+import com.maddyhome.idea.vim.impl.state.toMappingMode
+import com.maddyhome.idea.vim.newapi.vim
+import com.maddyhome.idea.vim.options.OptionAccessScope
+import eu.theblob42.idea.whichkey.config.MappingConfig
+import eu.theblob42.idea.whichkey.config.PopupConfig
+import eu.theblob42.idea.whichkey.model.Mapping
+import java.awt.event.KeyEvent
+import javax.swing.KeyStroke
+
+class WhichKeyActionListener : AnActionListener {
+    private var ignoreNextExecute = false
+
+    override fun beforeShortcutTriggered(
+        shortcut: Shortcut,
+        actions: MutableList<AnAction>,
+        dataContext: DataContext
+    ) {
+        if (shortcut !is KeyboardShortcut) {
+            return
+        }
+
+        val editor = dataContext.getData(CommonDataKeys.EDITOR) ?: return
+        val commandState = CommandState.getInstance(editor)
+        val mappingState = commandState.mappingState
+        val escapeKeyStroke = KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0)
+        val mappingMode = editor.vim.mode.toMappingMode()
+        val vimCurrentKeySequence = commandState.mappingState.keys.toList().ifEmpty { commandState.commandBuilder.keys.toList() }
+
+        // check if user pressed escape and we are in the middle of a sequence
+        // and ESC is not a possible next element in the sequence.
+        // If this is the case, treat it as 'cancel'
+        if (listOfNotNull(shortcut.firstKeyStroke, shortcut.secondKeyStroke) == listOf(escapeKeyStroke)
+            && MappingConfig.getNestedMappings(mappingMode, vimCurrentKeySequence).isNotEmpty()
+            && MappingConfig.getNestedMappings(mappingMode, vimCurrentKeySequence + escapeKeyStroke).isEmpty()
+        ) {
+            mappingState.resetMappingSequence()
+            return
+        }
+
+        // this event fires BEFORE handling by ideavim, so we need to construct the sequence ourselves
+        val typedKeySequence =
+            vimCurrentKeySequence + listOfNotNull(shortcut.firstKeyStroke, shortcut.secondKeyStroke)
+        // SPACE registers both as a shortcut as well as "typed space" event, so ignore the shortcut variant
+        if (typedKeySequence == listOf(KeyStroke.getKeyStroke(KeyEvent.VK_SPACE, 0))) {
+            return
+        }
+        processKeySequence(editor, typedKeySequence)
+    }
+
+    override fun afterEditorTyping(charTyped: Char, dataContext: DataContext) {
+        val editor = dataContext.getData(CommonDataKeys.EDITOR) ?: return
```

Comment:
```suggestion
        PopupConfig.hidePopup()
        val editor = dataContext.getData(CommonDataKeys.EDITOR) ?: return
```
Make sure the popup is hidden before any processing, otherwise we might end up with lasting popup windows. Especially when the entered key sequence deleted the current window (in this case `editor` is null)

Reply:
Just for reference I could easily reproduce this on my end with the following "window closing" mapping typed really quick (not working with the last window):
```
nnoremap <leader>wd    :action CloseAllEditors<CR>
```

Reply:
I wasn't able to reproduce it with that binding, but I had the same issue with my `lazyvim` based config when pressing `<leader>f` quickly. Fixed!